### PR TITLE
*: Push down where in newTableScan

### PR DESCRIPTION
**Pull request by zimulala**

### CODE REVIEW

**Comment by zimulala** [comment]
PTAL @hanfei19910905 @shenli @coocood

File: plan/refiner.go [hunk 1]
```diff
@@ -44,10 +45,9 @@ func refine(in Plan) error {
 	case *TableScan:
 		err = buildTableRange(x)
 	case *NewTableScan:
-		x.Ranges = []TableRange{{math.MinInt64, math.MaxInt64}}
-		return nil
+		err = buildNewTableRange(x)
```
Comment (coocood) [inline plan/refiner.go]: Get conditions from the parent `*Selection` of `*NewTableScan`, then build new table range.

Comment (zimulala) [inline plan/refiner.go]: Got it.

File: plan/refiner.go [hunk 2]
```diff
@@ -93,6 +94,33 @@ func buildTableRange(p *TableScan) error {
...
+func buildNewTableRange(p *NewTableScan) error {
+	// Convert selection plan's conditions to newTableScan plan's AccessConditions or FilterConditions.
+	var ok bool
+	var selection *Selection
+	for _, parent := range p.GetParents() {
+		if selection, ok = parent.(*Selection); ok {
+			break
+		}
+	}
+	if !ok || len(selection.Conditions) == 0 {
```
Comment (coocood) [inline plan/refiner.go]: 'selection == nil' is more readable than '!ok'.

File: executor/new_executor.go [hunk 3]
```diff
@@ -676,3 +681,123 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+	tbl *model.TableInfo) *tipb.Expr {
+	var tp tipb.ExprType
+	switch expr.FuncName.L {
+	case "<":
+		tp = tipb.ExprType_LT
+	case "<=":
+		tp = tipb.ExprType_LE
+	case "=":
+		tp = tipb.ExprType_EQ
+	case "!= ":
```
Comment (coocood) [inline executor/new_executor.go]: an extra space.

File: executor/new_executor.go [hunk 4]
```diff
@@ -676,3 +681,123 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+	case ">":
+		tp = tipb.ExprType_GT
+	case "<=>":
+		tp = tipb.ExprType_NullEQ
+	case "&&":
+		tp = tipb.ExprType_And
+	case "||":
+		tp = tipb.ExprType_Or
+	// It's the operation for unary operator.
+	case opcode.Ops[opcode.Not]:
```
Comment (coocood) [inline executor/new_executor.go]: Why not just use "!"?

File: executor/new_executor.go [hunk 5]
```diff
@@ -676,3 +681,123 @@ func (e *NewSortExec) Next() (*Row, error) {
...
+	return &tipb.Expr{
+		Tp:  tipb.ExprType_ColumnRef.Enum(),
+		Val: codec.EncodeInt(nil, id)}
+}
+
+func (b *executorBuilder) scalarFuncToPBExpr(client kv.Client, expr *expression.ScalarFunction,
+	tbl *model.TableInfo) *tipb.Expr {
+	var tp tipb.ExprType
+	switch expr.FuncName.L {
+	case "<":
```
Comment (hanfei19910905) [inline executor/new_executor.go]: opcode.Ops[opcode.LT] is better

File: plan/newplanbuilder.go [hunk 6]
```diff
@@ -230,14 +234,18 @@ func (b *planBuilder) buildProjection(src Plan, fields []*ast.SelectField, mappe
...
 			colTblName := field.WildCard.Table
 			for _, col := range src.GetSchema() {
-				if (dbName.L == "" || dbName.L == col.DBName.L) &&
-					(colTblName.L == "" || colTblName.L == col.TblName.L) {
-					newExpr := col.DeepCopy()
-					proj.Exprs = append(proj.Exprs, newExpr)
-					tblName = col.TblName
-					schemaCol := &expression.Column{FromID: col.FromID, TblName: tblName, ColName: col.ColName, RetType: newExpr.GetType()}
-					schema = append(schema, schemaCol)
+				if (dbName.L != "" && dbName.L != col.DBName.L) ||
```
Comment (hanfei19910905) [inline plan/newplanbuilder.go]: why do this ?

File: plan/new_plans.go [hunk 7]
```diff
@@ -93,9 +94,24 @@ type NewTableScan struct {
 	LimitCount *int64
 }
 
-func (ts *NewTableScan) attachCondition(conditions []expression.Expression) {
+func (ts *NewTableScan) attachConditions(conditions []expression.Expression) {
```
Comment (hanfei19910905) [inline plan/new_plans.go]: If you build range from selection, then attachConditions should be abandoned.

File: executor/new_builder.go [hunk 8]
```diff
@@ -140,6 +142,18 @@ func (b *executorBuilder) buildNewTableScan(v *plan.NewTableScan) Executor {
 			Columns:     v.Columns,
 			ranges:      v.Ranges,
 		}
+		var remained []expression.Expression
+		e.where, remained = b.newConditionExprToPBExpr(client, v.FilterConditions, v.Table)
```
Comment (hanfei19910905) [inline executor/new_builder.go]: If you build range from selection, then get filter conditions from parent selection.

File: executor/new_builder.go [hunk 9]
```diff
@@ -130,7 +132,7 @@ func (b *executorBuilder) buildNewTableScan(v *plan.NewTableScan) Executor {
 	supportDesc := client.SupportRequestType(kv.ReqTypeSelect, kv.ReqSubTypeDesc)
 	if !memDB && client.SupportRequestType(kv.ReqTypeSelect, 0) {
 		// TODO: support condition pushdown and union scan exec.
```
Comment (shenli) [inline executor/new_builder.go]: update this TODO

File: executor/new_builder.go (hunk 8 above)
Comment (zimulala) [inline executor/new_builder.go]: I do it in refiner file.

File: plan/range.go [hunk 10]
```diff
@@ -116,6 +117,94 @@ func (r *rangeBuilder) build(expr ast.ExprNode) []rangePoint {
 	return fullRange
 }
 
+func (r *rangeBuilder) newBuild(expr expression.Expression) []rangePoint {
+	switch x := expr.(type) {
+	case *expression.Column:
+		return r.buildFromColumn(x)
+	case *expression.ScalarFunction:
```
Comment (shenli) [inline plan/range.go]: Consider expression.Constant? @hanfei19910905

File: plan/refiner.go (hunk 2 above)
Comment (shenli) [inline plan/refiner.go]: Or rename ok to hasSelection.

File: plan/range.go (hunk 10 above)
Comment (hanfei19910905) [inline plan/range.go]: Yes, we need.

File: plan/newplanbuilder.go [hunk 11]
```diff
@@ -116,19 +117,21 @@ func extractColumn(expr expression.Expression, cols []*expression.Column) (resul
...
 		binop, ok := expr.(*expression.ScalarFunction)
-		eqStr, _ := opcode.Ops[opcode.EQ]
-		if ok && binop.FuncName.L == eqStr {
+		if ok && binop.FuncName.L == ast.EQ {
 			ln, lOK := binop.Args[0].(*expression.Column)
 			rn, rOK := binop.Args[1].(*expression.Column)
 			if lOK && rOK {
 				if left.GetSchema().GetIndex(ln) != -1 && right.GetSchema().GetIndex(rn) != -1 {
 					eqCond = append(eqCond, binop)
 					continue
```
Comment (zimulala) [inline plan/newplanbuilder.go]: I think four layers of nested isn't good.

File: plan/newplanbuilder.go (hunk 6 above)
Comment (zimulala) [inline plan/newplanbuilder.go]: I think four layers of nested isn't good.

File: plan/range.go (hunk 10 above)
Comment (coocood) [inline plan/range.go]: If the constant value is 0 or null, we should return empty range, otherwise return full range.

**Comment by zimulala** [comment]
PTAL @coocood @shenli @hanfei19910905

**Comment by coocood** [comment]
LGTM

File: plan/refiner.go [hunk 12]
```diff
@@ -191,3 +258,50 @@ func (c *conditionChecker) checkColumnExpr(expr ast.ExprNode) bool {
...
+	}
+	return false
+}
+
+func (c *conditionChecker) checkScalarFunction(scalar *expression.ScalarFunction) bool {
+	// TODO: Implement istrue, between, parentheses, patternin and patternlike.
+	switch scalar.FuncName.L {
+	case ast.OrOr:
+		return c.newCheck(scalar.Args[0]) && c.newCheck(scalar.Args[1])
+	case ast.AndAnd:
```
Comment (hanfei19910905) [inline plan/refiner.go]: case ast.OrOr, ast.AndAnd:

**Comment by hanfei19910905** [comment]
rest LGTM

**Pull request merged**